### PR TITLE
Leadership Pinning Client API Server.

### PR DESCRIPTION
**Pull request by manadart**
## Description of change

This patch threads the Pinner interface through the API server context and adds a new client API for pinning and unpinning leadership.

## QA steps

New unit tests; currently no functionality materialised.

## Documentation changes

None.

## Bug reference

N/A

### CODE REVIEW

**Comment by manadart** [comment]
[Expose Leadership Pinning in the API Server](https://trello.com/c/u00f0dRG/643-expose-leadership-pinning-in-the-api-server)

File: apiserver/facade/registry_test.go [hunk 1]
```diff
@@ -270,11 +270,11 @@ func assertRegisterFlag(c *gc.C, registry *facade.Registry, name string, version
 	c.Assert(err, gc.IsNil)
 }
 
-func testFacade(facade.Context) (facade.Facade, error) {
+func testFacade(_ facade.Context) (facade.Facade, error) {
```
Comment (manadart) [inline apiserver/facade/registry_test.go]: Drive-by appeasements of the linter.

**Comment by mitechie** [comment]
Looks like a legit failure 


    # github.com/juju/juju/apiserver/facades/client/charms_test 
    [github.com/juju/juju/apiserver/facades/client/charms.test]
    apiserver/facades/client/charms/client_test.go:62:52: cannot use charmsSuiteContext literal (type 
    *charmsSuiteContext) as type facade.Context in argument to charms.NewFacade:
    	*charmsSuiteContext does not implement facade.Context (missing LeadershipPinner method)

**Comment by manadart** [comment]
Yep; I will sort it tomorrow.

File: apiserver/leadership.go [hunk 2]
```diff
@@ -18,18 +18,18 @@ type leadershipChecker struct {
 }
 
 // LeadershipCheck is part of the leadership.Checker interface.
-func (m leadershipChecker) LeadershipCheck(applicationname, unitName string) leadership.Token {
-	token := m.checker.Token(applicationname, unitName)
+func (m leadershipChecker) LeadershipCheck(applicationName, unitName string) leadership.Token {
+	token := m.checker.Token(applicationName, unitName)
 	return leadershipToken{
-		applicationname: applicationname,
+		applicationName: applicationName,
```
Comment (ExternalReality) [inline apiserver/leadership.go]: +1 Leaving code better than you found it.

**Comment by manadart** [comment]
!!build!!

**Comment by manadart** [comment]
$$merge$$
